File: slack-poster/src/main/java/dev/number6/slackposter/ChannelSummaryImage.java
package dev.number6.slackposter;

import com.amazonaws.services.dynamodbv2.model.AttributeValue;

import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public class ChannelSummaryImage {

    public static final int FINAL_UPDATE_MINIMUM_VERSION = 4;
    public static final String CHANNEL_NAME_KEY = "channelName";
    public static final String COMPREHENSION_DATE_KEY = "comprehensionDate";
    public static final String VERSION_KEY = "version";
    public static final String SENTIMENT_TOTALS_KEY = "sentimentTotals";
    public static final String ENTITY_TOTALS_KEY = "entityTotals";
    public static final String KEYPHRASES_TOTALS_KEY = "keyPhrasesTotals";
    public static final String SENTIMENT_SCORE_TOTALS = "sentimentScoreTotals";

    private Map<String, AttributeValue> vals;

    public ChannelSummaryImage(Map<String, AttributeValue> vals) {
        this.vals = vals;
    }

    public boolean hasFinalUpdate() {
        return Integer.parseInt(vals.get(VERSION_KEY).getN()) >= FINAL_UPDATE_MINIMUM_VERSION;
    }

    public Map<String, Double> getSentimentScoreTotals() {
        Map<String, AttributeValue> sentimentScoreTotals = vals.get(SENTIMENT_SCORE_TOTALS).getM();
        return sentimentScoreTotals.entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey,
                        e -> Double.parseDouble(e.getValue().getN())));
    }

    public Map<String, Integer> getSentimentTotals() {
        Map<String, AttributeValue> sentimentTotals = vals.get(SENTIMENT_TOTALS_KEY).getM();
        return sentimentTotals.entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey,
                        e -> Integer.parseInt(e.getValue().getN())));
    }

    public Map<String, Integer> getKeyPhrasesTotals() {
        Map<String, AttributeValue> keyPhrasesTotals = vals.get(KEYPHRASES_TOTALS_KEY).getM();
        return keyPhrasesTotals.keySet().stream()
                .collect(Collectors.toMap(Function.identity(),
                        e -> Integer.parseInt(keyPhrasesTotals.get(e).getN())));
    }

    public String getChannelName() {
        return vals.get(CHANNEL_NAME_KEY).getS();
    }

    public String getComprehensionDate() {
        return vals.get(COMPREHENSION_DATE_KEY).getS();
    }

    public Map<String, Map<String, Integer>> getEntityTotals() {
        Map<String, AttributeValue> entityTotals = vals.get(ENTITY_TOTALS_KEY).getM();
        return entityTotals.entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey,
                        e -> e.getValue().getM().entrySet().stream()
                                .collect(Collectors.toMap(Map.Entry::getKey,
                                        e2 -> Integer.parseInt(e2.getValue().getN())))));
    }

    @Override
    public String toString() {
        return "ChannelSummaryImage{" +
                "vals=" + vals +
                '}';
    }
}

File: slack-poster/src/main/kotlin/dev/number6/slackposter/ChannelSummaryImage.kt
package dev.number6.slackposter

import com.amazonaws.services.dynamodbv2.model.AttributeValue

class ChannelSummaryImage(private val vals: Map<String, AttributeValue>) {
    fun hasFinalUpdate(): Boolean {
        return valNamed(VERSION_KEY).n.toInt() >= FINAL_UPDATE_MINIMUM_VERSION
    }

    val sentimentScoreTotals: Map<String, Double> by lazy {
        valNamed(SENTIMENT_SCORE_TOTALS).m
                .mapKeys { e -> e.key }
                .mapValues { e -> e.value.n.toDouble() }
    }

    val sentimentTotals: Map<String, Int> by lazy {
        valNamed(SENTIMENT_TOTALS_KEY).m
                .mapKeys { e -> e.key }
                .mapValues { e -> e.value.n.toInt() }
    }

    val keyPhrasesTotals: Map<String, Int> by lazy {
        valNamed(KEYPHRASES_TOTALS_KEY).m
                .mapKeys { e -> e.key }
                .mapValues { e -> e.value.n.toInt() }
    }

    val channelName: String by lazy { valNamed(CHANNEL_NAME_KEY).s }

    val comprehensionDate: String by lazy { valNamed(COMPREHENSION_DATE_KEY).s }

    val entityTotals: Map<String, Map<String, Int>> by lazy {
        val entityTotals = valNamed(ENTITY_TOTALS_KEY).m

        entityTotals.mapKeys { e -> e.key }
                .mapValues { e -> e.value.m.mapKeys { e1 -> e1.key }.mapValues { e1 -> e1.value.n.toInt() } }
    }

    private fun valNamed(key: String): AttributeValue {
        return vals[key] ?: error("Cannot find $key in $vals")
    }

    override fun toString(): String {
        return "ChannelSummaryImage{" +
                "vals=" + vals +
                '}'
    }

    companion object {
        const val FINAL_UPDATE_MINIMUM_VERSION = 4
        const val CHANNEL_NAME_KEY = "channelName"
        const val COMPREHENSION_DATE_KEY = "comprehensionDate"
        const val VERSION_KEY = "version"
        const val SENTIMENT_TOTALS_KEY = "sentimentTotals"
        const val ENTITY_TOTALS_KEY = "entityTotals"
        const val KEYPHRASES_TOTALS_KEY = "keyPhrasesTotals"
        const val SENTIMENT_SCORE_TOTALS = "sentimentScoreTotals"
    }
}